connector/connect/server/src/main/scala/org/apache/spark/sql/connect/utils/ErrorUtils.scala

```diff
@@ -181,7 +182,12 @@ private[connect] object ErrorUtils extends Logging {
         }
         val errorClass = e.getErrorClass
         if (errorClass != null && errorClass.nonEmpty) {
-          errorInfo.putMetadata("errorClass", errorClass)
+          val messageParameters = JsonMethods.compact(
+            JsonMethods.render(map2jvalue(e.getMessageParameters.asScala.toMap)))
+          if (messageParameters.length <= maxMetadataSize) {
+            errorInfo.putMetadata("errorClass", errorClass)
+            errorInfo.putMetadata("messageParameters", messageParameters)
```

Comment:
@HyukjinKwon please review this. I'm not sure if this is correct, as I don't see we pass the error context here (e.g. SQL string index).

Reply:
Without the error context you restore an error but without message parameter it is impossible, and just passing the error class is useless.

Reply:
Constructing exceptions based on error class and message parameters only works well for Scala client if I understand correctly? For Python client, it only recognizes a very limited set of error classes. https://github.com/databricks/runtime/blob/master/python/pyspark/errors/error_classes.py#L1114

Reply:
Another issue for constructing exception messages based on error class and message parameters on the client side is the multi version support because for the same error classes in different versions of clients, constructing error messages may require different message parameters

Reply:
> only works well for Scala client if I understand correctly? For Python client, it only recognizes a very limited set of error classes.

This only means that we should keep the python client in sync. Not more.

> constructing error messages may require different message parameters

The issue might be solved by sending old and new parameters together but if you send nothing, you don't have any chance.

Reply:
Make sense. Though the message parameters in error info metadata won't be used in most cases because of error enrichment. Also, we won't get error class and message parameters if  message parameters is too large

Reply:
Then why shouldn't we send the message parameters during error enrichment?

Reply:
We do? https://github.com/apache/spark/blob/0de70c4cb0b323b9802b3c8db61af042cc470742/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/utils/ErrorUtils.scala#L133

Reply:
@MaxGekk why do we need this fix then?

Reply:
In the PR https://github.com/apache/spark/pull/44464, I require all spark exceptions must have an error class (cannot build from just a text message). Found a few test failures at the client connect while creating a `SparkThrowable` in `GrpcExceptionConverter` like at:
```scala
    errorConstructor(params =>
      new SparkNumberFormatException(
        params.errorClass.orNull,
        params.messageParameters,
        params.queryContext)),
```

`SparkNumberFormatException`'s constructor fails because it cannot fill in message parameters since we don't transfer them (the map is empty).

Reply:
For example https://github.com/MaxGekk/spark/runs/19914247939, the failed test `ClientE2ETestSuite.cause exception - false`:
```
sbt.ForkMain$ForkError: org.scalatest.exceptions.TestFailedException: Expected exception org.apache.spark.SparkUpgradeException to be thrown, but org.apache.spark.SparkException was thrown
...
Caused by: sbt.ForkMain$ForkError: org.apache.spark.SparkException: [INTERNAL_ERROR] Undefined error message parameter for error class: 'INCONSISTENT_BEHAVIOR_CROSS_VERSION.PARSE_DATETIME_BY_NEW_PARSER'. Parameters: Map() SQLSTATE: XX000
	at org.apache.spark.SparkException$.internalError(SparkException.scala:92)
	at org.apache.spark.SparkException$.internalError(SparkException.scala:96)
```

How should I create `SparkUpgradeException` if its constructor requires an error class and **correct** message parameters?

Reply:
Let me make a followup to match Python side for now.

Reply:
https://github.com/apache/spark/pull/44528